### PR TITLE
[CALCITE-4661]Add MODE aggregate function(NobiGo)

### PR DESCRIPTION
**Pull request by NobiGo**
Add MODE functions in the operator table which returns the most frequent input value.

### CODE REVIEW

**Comment by zinking** [comment]
LGTM

**Comment by a-rafay** [comment]
Should we add execution tests for this?

File: core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java [hunk 1]
```diff
@@ -1241,6 +1243,81 @@ public static Expression multiplyDivide(Expression e, BigDecimal multiplier,
...
+                  Expressions.constant(0, Long.class))));
+    }
+
+    @Override protected void implementNotNullAdd(AggContext info,
+        AggAddContext add) {
+      Expression args = add.arguments().get(0);
+
+      Expression accResult = add.accumulator().get(0);
+      Expression accMap = add.accumulator().get(1);
+      Expression accLong = add.accumulator().get(2);
```
Comment (liyafan82) [inline core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java:1273]: Maybe better variable names could be used here, and some comments could be added to improve readability.

File: core/src/test/java/org/apache/calcite/sql/test/SqlOperatorBaseTest.java [hunk 2]
```diff
@@ -7603,6 +7603,39 @@ void checkArrayConcatAggFuncFails(SqlTester t) {
     tester.checkAgg("intersection(x)", values3, "[0, 1, 1]", 0);
   }
 
+  @Test void testModeFunc() {
+    tester.setFor(SqlStdOperatorTable.MODE, VM_EXPAND);
+    tester.checkFails(
+        "mode(^*^)", "Unknown identifier '\\*'", false);
+    strictTester.checkFails(
+        "^mode()^",
+        "Invalid number of arguments to function 'MODE'. Was expecting 1 arguments",
```
Comment (liyafan82) [inline core/src/test/java/org/apache/calcite/sql/test/SqlOperatorBaseTest.java:7612]: nit: 1 argument?

Comment (NobiGo) [reply core/src/test/java/org/apache/calcite/sql/test/SqlOperatorBaseTest.java:7612]: Yes. First thanks for your review. You can find related info [CALCITE-4661](https://issues.apache.org/jira/projects/CALCITE/issues/CALCITE-4661). The MODE function in calcite doesn't the same as PostgreSQL, Just a normal agg function.

File: core/src/test/java/org/apache/calcite/test/SqlToRelConverterTest.java [hunk 3]
```diff
@@ -3843,6 +3843,33 @@ private Tester getExtendedTester() {
     sql(sql).ok();
   }
 
+  /** Test case for
+   * <a href="https://issues.apache.org/jira/browse/CALCITE-4661">[CALCITE-4661]
+   * Add MODE aggregate function</a>. */
+  @Test public void testModeFunction() {
```
Comment (liyafan82) [inline core/src/test/java/org/apache/calcite/test/SqlToRelConverterTest.java:3849]: remove "public"?

File: core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java (hunk 1 above)
Comment (NobiGo) [reply core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java:1273]: Thank you very much for your advice. I will make some improvements

**Comment by NobiGo** [comment]
> Should we add execution tests for this?

Thank you very much for your advice. I will make some improvements.

File: core/src/test/java/org/apache/calcite/sql/test/SqlOperatorBaseTest.java (hunk 2 above)
Comment (liyafan82) [reply core/src/test/java/org/apache/calcite/sql/test/SqlOperatorBaseTest.java:7612]: I see. Thanks for the explanation.
I mean the message should be "1 argument"? Anyway, this is a minor issue, so you may choose to accept it or not.

Comment (NobiGo) [reply core/src/test/java/org/apache/calcite/sql/test/SqlOperatorBaseTest.java:7612]: Sorry. I get it.

**Comment by NobiGo** [comment]
> Should we add execution tests for this?

done

**Comment by amaliujia** [comment]
Sorry I have been busy recently so failed to keep my promise of checking this PR. I will walk through the JIRA content and go back to this PR soon.

File: core/src/main/java/org/apache/calcite/sql/fun/SqlStdOperatorTable.java [hunk 4]
```diff
@@ -952,6 +952,16 @@
    */
   public static final SqlAggFunction COUNT = new SqlCountAggFunction("COUNT");
 
+  /**
+   * <code>MODE</code> aggregate function.
+   */
+  public static final SqlAggFunction MODE =
+      SqlBasicAggFunction
+          .create("MODE", SqlKind.MODE,
+              ReturnTypes.ARG0_NULLABLE_IF_EMPTY, OperandTypes.ANY_NOT_NULL)
```
Comment (amaliujia) [inline core/src/main/java/org/apache/calcite/sql/fun/SqlStdOperatorTable.java:961]: In the JIRA discussion, I see that MODE allows NULL but if NULL is the most common value, return the second most common value instead. Why you choose to does not allow now in type?

File: site/_docs/reference.md [hunk 5]
```diff
@@ -1824,33 +1824,34 @@ and `LISTAGG`).
```
Comment (amaliujia) [inline site/_docs/reference.md:1824]: Can you revert the change in reference.md and then only add MODE documentation? If there is a need to re-order and fix existing function documentation, that should be done by another PR with its own JIRA. Let's keep this PR only for support MODE funciton.

Comment (NobiGo) [reply site/_docs/reference.md:1824]: This change according to the JIRA code review:

- In reference.md, sort the table of aggregate functions into alphabetical order.

File: core/src/main/java/org/apache/calcite/sql/fun/SqlStdOperatorTable.java (hunk 4 above)
Comment (NobiGo) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlStdOperatorTable.java:961]: Yes. You are right. But This parameter means cant directly MODE(NULL). When using MODE(ColumnName) if the column value most common value is NULL, will return the second most common value insteed.

File: site/_docs/reference.md (hunk 5 above)
Comment (amaliujia) [reply site/_docs/reference.md:1824]: Ah I see. Ok we can do it once then in the same PR.

File: core/src/main/java/org/apache/calcite/sql/fun/SqlStdOperatorTable.java (hunk 4 above)
Comment (amaliujia) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlStdOperatorTable.java:961]: Ah thanks for the clarification. Let me double check `OperandTypes.ANY_NOT_NULL`

Comment (NobiGo) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlStdOperatorTable.java:961]: If no other problem about this, Please let me known or "Resolve conversation". Thank you.

File: core/src/test/resources/sql/agg.iq [hunk 6]
```diff
@@ -2990,4 +2990,146 @@ from (select 1 as a, 2 as b, 3 as c);
...
+
+# mode function applied to filter and input value is NULL.
+select MODE(gender)
+from emp
+where deptno > 60;
++--------+
+| EXPR$0 |
++--------+
+|        |
++--------+
```
Comment (amaliujia) [inline core/src/test/resources/sql/agg.iq:3040]: Interesting. So if the input is only NULL so nothing will be returned?

File: core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java [hunk 7]
```diff
@@ -1241,6 +1243,83 @@ public static Expression multiplyDivide(Expression e, BigDecimal multiplier,
...
+      // acc[2] = Long.valueOf(0);
+      reset.currentBlock().add(
+          Expressions.statement(
+              Expressions.assign(reset.accumulator().get(2),
+                  Expressions.constant(0, Long.class))));
+    }
+
+    @Override protected void implementNotNullAdd(AggContext info,
+        AggAddContext add) {
+      Expression currentArg = add.arguments().get(0);
```
Comment (amaliujia) [inline core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java:1269]: Nit: this will be very useful if you can also add equivalent code/pesudo code into comment.

**Review by amaliujia — COMMENTED** [review]
Overall LGTM

File: core/src/test/resources/sql/agg.iq (hunk 6 above)
Comment (NobiGo) [reply core/src/test/resources/sql/agg.iq:3040]: Yes. This result keeps consistent with PostgreSQL.

File: core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java (hunk 7 above)
Comment (NobiGo) [reply core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java:1269]: arguments method just can get the column value. So I don't add a comment

File: core/src/test/resources/sql/agg.iq (hunk 6 above)
Comment (NobiGo) [reply core/src/test/resources/sql/agg.iq:3040]: @amaliujia If you have any other advice,  please let me know. Then we can make some progress.
